Logic for nested HttpRequestMethodConfig Values

bradleypettit opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Fixes #187. Adds logic to the compare_condition function to suit when the 'Values' key is nested within 'HttpRequestMethodConfig'. Otherwise trying to access 'Values' directly throws "KeyError: 'Values'".
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
elbv2.py
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
As per [Boto docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.create_rule), when condition Field is equal to 'http-request-method', the 'Values' list should then be nested within 'HttpRequestMethodConfig' dict.
<!--- Paste verbatim command output below, e.g. before and after your change -->

bradleypettit commented:
I've found that the issue I was experiencing with HttpRequestMethodConfig applies to all other field categories with the exception of SourceIpConfig. SourceIpConfig already had a similar condition to the ones I have added for other fields.

Updated the linked issue with additional detail and updated this PR to resolve issue for all possible condition field options.

I believe this PR also fixes https://github.com/ansible-collections/community.aws/issues/117

bradleypettit commented:
Found that while adding support for using HostHeaderConfig or PathPatternConfig I'd inadvertantly impacted cases when Values was defined directly instead.

This is specific to the Boto3 behaviour where Conditions['Values'] ([Docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.create_rule)) is defined as "_The condition value. Specify only when Field is host-header or path-pattern . Alternatively, to specify multiple host names or multiple path patterns, use HostHeaderConfig or PathPatternConfig ._"

For example when _compare_condition is running against an ALB with an existing host-header field, the current_condition object is represented as:
```
{
'Field': 'host-header',
'HostHeaderConfig': { 'Values': ['host.1'] }
'Values': ['host.1']
}
```
As a result, my original update was then always trying to compare HostHeaderConfig['Values'], however it is expected that in ansible you can use either Values or HostHeaderConfig.

While I've used host-header (HostHeaderConfig) primarily in the above example - the same applies to path-pattern (PathPatternConfig).

To cater for both possible methods of configuration, I've updated the conditions to only use HostHeaderConfig or PathPatternConfig for comparison if both current_condition and condition contain the Config dict, otherwise fallback to the pre-existing Values compare.

goneri commented:
Hi @bradleypettit,

The patch indeed seems to fix https://github.com/ansible-collections/community.aws/issues/117 too. I need more time to review it completely.
Thank you for your contribution! My goal is to get it merged quickly now.

goneri commented:
I would like to land https://github.com/ansible-collections/community.aws/pull/347 with this PR, there is a couple of CI-related issues to address first.

goneri commented:
This feature depends on boto3 June 13, 2019:

```
AWS::ElasticLoadBalancingV2::ListenerRule
In the RuleCondition property type:
Use the HostHeaderConfig property to specify information for a host header condition.
Use the HttpHeaderConfig property to specify information for an HTTP header condition.
Use the HttpRequestMethodConfig property to specify information for an HTTP method condition.
Use the PathPatternConfig property to specify information for a path pattern condition.
Use the QueryStringConfig property to specify information for a query string condition.
Use the SourceIpConfig property to specify information for a source IP condition.
```

See: https://forums.aws.amazon.com/thread.jspa?messageID=897663

goneri commented on changelogs/fragments/188-httprequestmethodconfig-keyerror.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+bugfixes:
+  - module_utils/elbv2 - add logic to compare_rules to suit Values list nested within dicts unique to each field type. 
+    Fixes issue (https://github.com/ansible-collections/amazon.aws/issues/187)
```
Can you add a message saying this also requires boto3 1.16.57 or greater.

goneri commented on plugins/module_utils/elbv2.py:
```diff
@@ -719,11 +719,46 @@ def _compare_condition(self, current_conditions, condition):
         condition_found = False
 
         for current_condition in current_conditions:
-            if current_condition.get('SourceIpConfig'):
```
Can you raise an error if the boto3 version is 1.16.57 <. I think this would save a lot of painful troubleshooting for the users.

For instance:  https://github.com/ansible-collections/community.aws/blob/main/plugins/modules/ec2_launch_template.py#L689-L690

bradleypettit replied:
Thanks for the reference. I've added a boto3 version check when the ElasticLoadBalancerV2 class is initialised. Let me know if it'd be better placed elsewhere.

goneri commented:
waiting_on_contributor

bradleypettit commented:
Thanks for your help @goneri, I've made changes as you suggested.

Snippet of using community.aws.elb_application_lb task with an older and current boto3 version below. 

**boto3==1.16.56**
```
$ pip3 list | grep boto3
boto3                  1.16.56
$ ansible-playbook aws_alb_provision.yml
...
TASK [ALB Rule - http-header] **********************************************************************************************************************************************
failed: [localhost] (item=0) => {"ansible_loop_var": "item", "changed": false, "item": 0, "msg": "elbv2 requires boto3 >= 1.16.57"}
failed: [localhost] (item=1) => {"ansible_loop_var": "item", "changed": false, "item": 1, "msg": "elbv2 requires boto3 >= 1.16.57"}
...
```
**boto3==1.16.59**
```
$ pip3 list | grep boto3
boto3                  1.16.59
$ ansible-playbook aws_alb_provision.yml
...
TASK [ALB Rule - http-header] **********************************************************************************************************************************************
changed: [localhost] => (item=0)
changed: [localhost] => (item=1)
...
```

goneri commented:
Thank you @bradleypettit for your contribution!
